docs/docsite/rst/dev_guide/developing_module_utilities.rst

```diff
@@ -4,84 +4,63 @@
+``basic.py`` module utility provides the main entry point for accessing the
+Ansible library, and all Python Ansible modules must import something from
+``ansible.module_utils``. A common option is to import ``AnsibleModule``::
 
   from ansible.module_utils.basic import AnsibleModule
 
+The ``ansible.module_utils`` namespace is not a plain Python package: it is
+constructed dynamically for each task invocation, by extracting imports and
+resolving those matching the namespace against a search path derived from the
+active configuration.
+
 If you need to share Python code between some of your own local modules, you can use Ansible's ``module_utils`` directories for this. When you run ``ansible-playbook``, Ansible will merge any files in the local ``module_utils`` directory into the ``ansible.module_utils`` namespace. For example, if you have your own custom modules that import a ``my_shared_code`` library, you can place that into a ``./module_utils/my_shared_code.py`` file in the root location where your playbook lives, and then import it in your modules like so::
 
   from ansible.module_utils.my_shared_code import MySharedCodeClient
 
 Your custom ``module_utils`` directories can live in the root directory of your playbook, or in the individual role directories, or in the directories specified by the ``ANSIBLE_MODULE_UTILS`` configuration setting.
 
-Ansible ships with the following list of ``module_utils`` files. The module utility source code lives in the ``./lib/ansible/module_utils`` directory under your main Ansible path - for more details on any specific module utility, please see the source code.
+Naming and finding module utilities
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Ansible uses namespacing to organize module utilities. We store generic
+utilities in the ``common`` subdirectory or in the root directory. For example,
+the shared code for parsing URLs is in ``lib/ansible/module_utils/urls.py``.
+Code files with a more specific purpose live in more specific
+directories. For example, modules related to EMC live in ``lib/ansible/modules/storage/emc/``
+and the module utilities related to EMC live in ``lib/ansible/module_utils/storage/emc/``.
+Following this pattern with your own module utilities makes everything easy to find and use.
+
+.. _standard_mod_utils:
+
+Standard module utilities
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Ansible ships with a comprehensive library of ``module_utils`` files.
+You can find the module
+utility source code in the ``./lib/ansible/module_utils`` directory under
+your main Ansible path. Most module utilities have names that describe their
+functions - for example, ``openstack.py`` contains utilities for modules that work with Openstack instances. We've described the most widely-used utilities below. For more details on any specific module utility,
+please see the `source code <https://github.com/ansible/ansible/tree/devel/lib/ansible/module_utils>`_.
 
 .. include:: shared_snippets/licensing.txt
 
```

Comment:
Please add
`api.py`  - This module adds shared support for generic API modules
`common/dict_transformations.py`
`common/file.py`
`distro/_distro.py` - bundled copy of distro?
`json_utils.py`
``

@abadger @nitzmahone 
What else from `module_utils/common/` do we need to list here?
`parsing/convert_bool.py` is that worth mentioning, or do you access it via `basic`?
`_text.py`?

Reply:
@sdoran this might be your area of expertise now.

Reply:
addressed @gundalow's suggestions in 584f57483bbb5501037da59219fecafb820bbbc2 - will follow up with @samdoran